janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java

```diff
@@ -989,6 +988,76 @@ public void testConditionalIndexing() {
+            assertEquals("Or(JanusGraphStep([],[name.eq(bob)]),JanusGraphStep([],[prop.eq(100)]))", mMixedOr.getName());
+            assertTrue(mMixedOr.getDuration(TimeUnit.MICROSECONDS) > 0);
+            assertEquals(2, mMixedOr.getNested().size());
+            Metrics nested = (Metrics) mMixedOr.getNested().toArray()[0];
+            assertEquals(QueryProfiler.CONSTRUCT_GRAPH_CENTRIC_QUERY, nested.getName());
+            assertTrue(nested.getDuration(TimeUnit.MICROSECONDS) > 0);
+            nested = (Metrics) mMixedOr.getNested().toArray()[1];
+            assertEquals(QueryProfiler.GRAPH_CENTRIC_QUERY, nested.getName());
+            assertTrue(nested.getDuration(TimeUnit.MICROSECONDS) > 0);
+            Map<String, String> annotations = new HashMap() {{
+                put("condition", "((name = bob) OR (prop = 100))");
+                put("orders", "[]");
+                put("isFitted", "false");
+                put("isOrdered", "true");
+                put("query", "[((name = bob) OR (prop = 100))](2000):mixed");
+                put("index", "mixed");
+                put("index_impl", "search");
+            }};
+            assertEquals(annotations, nested.getAnnotations());
+        }
+
+        // satisfied by a single graph-centric query which is satisfied by a single mixed index query
+        newTx();
+        Metrics mMixedAnd = tx.traversal().V().has("name", "bob").has("prop", 100)
+            .profile().next().getMetrics(0);
+        assertEquals("JanusGraphStep([],[name.eq(bob), prop.eq(100)])", mMixedAnd.getName());
+        assertTrue(mMixedAnd.getDuration(TimeUnit.MICROSECONDS) > 0);
+        assertEquals(3, mMixedAnd.getNested().size());
+        Metrics nested = (Metrics) mMixedAnd.getNested().toArray()[0];
+        assertEquals(QueryProfiler.CONSTRUCT_GRAPH_CENTRIC_QUERY, nested.getName());
+        assertTrue(nested.getDuration(TimeUnit.MICROSECONDS) > 0);
+        nested = (Metrics) mMixedAnd.getNested().toArray()[1];
+        assertEquals(QueryProfiler.CONSTRUCT_GRAPH_CENTRIC_QUERY, nested.getName());
+        assertTrue(nested.getDuration(TimeUnit.MICROSECONDS) > 0);
+        nested = (Metrics) mMixedAnd.getNested().toArray()[2];
+        assertEquals(QueryProfiler.GRAPH_CENTRIC_QUERY, nested.getName());
+        assertTrue(nested.getDuration(TimeUnit.MICROSECONDS) > 0);
+        String query;
+        if (readConfig.get(INDEX_BACKEND, INDEX).equals("solr")) {
+            query = "[(name_s = bob AND prop_i = 100)](1000):mixed";
+        } else {
+            query = "[(name = bob AND prop = 100)](1000):mixed";
+        }
+        Map<String, String> annotations = new HashMap() {{
+            put("condition", "(name = bob AND prop = 100)");
+            put("orders", "[]");
+            put("isFitted", "true");
+            put("isOrdered", "true");
+            put("query", query);
+            put("index", "mixed");
+            put("index_impl", "search");
```

Comment:
Do you like to test two different index_impl's at same time.

Reply:
That sounds interesting. Do you know if we have any existing test case that involves more than one index backend?

Reply:
I have no idea. I can check for it.

Reply:
We don't have any for test cases for it, as far as i can see.

Reply:
https://github.com/JanusGraph/janusgraph/blob/master/janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticsearchIndexTest.java#L409
 I found something.

Reply:
I tried it. Essentially, a gremlin query involving two index queries against two index backends is the same as a gremlin query involving two index queries against the same index backend, from the profiling perspective.

I think it would be good to have multiple index backends involved anyway... Will do that in the next PR (to fix AND/OR flattening issue). Right now the annotations are corrupted when there are multiple backend queries.

Reply:
Created https://github.com/JanusGraph/janusgraph/pull/2377 to support multiple index backends in JanusGraphIndexTest.